gcs: take less of forever to connect.

mlyle opened the pull request:
Also update the connection status object more frequently so we're less
likely to spuriously lose connection from little blips / lost messages.

fixes #582 

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/d-ronin/dronin/524)

<!-- Reviewable:end -->

mlyle committed:
gcs: uploader: trigger immediate boot when we don't want the loader

Makes things connect about 4-5s faster.

mlyle commented:
Verified on Quanton and RevoMini; did lots of cycles of rescue and just plugging in board.  Boots / connects much faster and happier. 

Didn't get any of the GCS crashes I'm used to with bootloader scenario, so it's possible I've even fixed regression with how these states / signals are managed.

tracernz commented:
Tested well on Revo and Sparky1. Often causes a hang with CC3D @ [hidapi_linux.c:1255](https://github.com/d-ronin/dRonin/blob/40216c88444cabee4ac7c3dfe2e6700ee6398b60/ground/gcs/src/plugins/rawhid/hidapi/hidapi_linux.c#L1255). BT if it helps:

```

Thread 22 (Thread 0x7fff819fe700 (LWP 22717)):
#0  pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
No locals.
#1  0x00007ffff618b7fb in QWaitCondition::wait(QMutex*, unsigned long) () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#2  0x00007ffff4f31c92 in ?? () from /home/mike/Dev/dronin/tools/Qt5.5.0/5.5/gcc_64/lib/libQt5Quick.so.5
No symbol table info available.
#3  0x00007ffff4f320b0 in ?? () from /home/mike/Dev/dronin/tools/Qt5.5.0/5.5/gcc_64/lib/libQt5Quick.so.5
No symbol table info available.
#4  0x00007ffff618a7df in ?? () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#5  0x00007ffff5ed1182 in start_thread (arg=0x7fff819fe700) at pthread_create.c:312
        __res = <optimized out>
        pd = 0x7fff819fe700
        now = <optimized out>
        unwind_buf = {cancel_jmp_buf = {{jmp_buf = {140735368128256, 591285074036167023, 1, 0, 140735368128960, 140735368128256, -591192439064456849, -591306914559651473}, mask_was_saved = 0}}, priv = {pad = {0x0, 0x0, 0x0, 0x0}, data = {prev = 0x0, cleanup = 0x0, canceltype = 0}}}
        not_first_call = <optimized out>
        pagesize_m1 = <optimized out>
        sp = <optimized out>
        freesize = <optimized out>
        __PRETTY_FUNCTION__ = "start_thread"
#6  0x00007ffff56e447d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
No locals.

Thread 21 (Thread 0x7fff82dff700 (LWP 22716)):
#0  pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
No locals.
#1  0x00007ffff618b7fb in QWaitCondition::wait(QMutex*, unsigned long) () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#2  0x00007ffff4f31c92 in ?? () from /home/mike/Dev/dronin/tools/Qt5.5.0/5.5/gcc_64/lib/libQt5Quick.so.5
No symbol table info available.
#3  0x00007ffff4f320b0 in ?? () from /home/mike/Dev/dronin/tools/Qt5.5.0/5.5/gcc_64/lib/libQt5Quick.so.5
No symbol table info available.
#4  0x00007ffff618a7df in ?? () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#5  0x00007ffff5ed1182 in start_thread (arg=0x7fff82dff700) at pthread_create.c:312
        __res = <optimized out>
        pd = 0x7fff82dff700
        now = <optimized out>
        unwind_buf = {cancel_jmp_buf = {{jmp_buf = {140735389103872, 591285074036167023, 1, 0, 140735389104576, 140735389103872, -591186393361116817, -591306914559651473}, mask_was_saved = 0}}, priv = {pad = {0x0, 0x0, 0x0, 0x0}, data = {prev = 0x0, cleanup = 0x0, canceltype = 0}}}
        not_first_call = <optimized out>
        pagesize_m1 = <optimized out>
        sp = <optimized out>
        freesize = <optimized out>
        __PRETTY_FUNCTION__ = "start_thread"
#6  0x00007ffff56e447d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
No locals.

Thread 19 (Thread 0x7fffa1310700 (LWP 22713)):
#0  0x00007ffff56d712d in poll () at ../sysdeps/unix/syscall-template.S:81
No locals.
#1  0x00007ffff39e3fe4 in ?? () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
No symbol table info available.
#2  0x00007ffff39e40ec in g_main_context_iteration () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
No symbol table info available.
#3  0x00007ffff63cdae7 in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#4  0x00007ffff6376252 in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#5  0x00007ffff618596c in QThread::exec() () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#6  0x00007ffff618a7df in ?? () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#7  0x00007ffff5ed1182 in start_thread (arg=0x7fffa1310700) at pthread_create.c:312
        __res = <optimized out>
        pd = 0x7fffa1310700
        now = <optimized out>
        unwind_buf = {cancel_jmp_buf = {{jmp_buf = {140735897732864, 591285074036167023, 1, 0, 140735897733568, 140735897732864, -591122990517022353, -591306914559651473}, mask_was_saved = 0}}, priv = {pad = {0x0, 0x0, 0x0, 0x0}, data = {prev = 0x0, cleanup = 0x0, canceltype = 0}}}
        not_first_call = <optimized out>
        pagesize_m1 = <optimized out>
        sp = <optimized out>
        freesize = <optimized out>
        __PRETTY_FUNCTION__ = "start_thread"
#8  0x00007ffff56e447d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
No locals.

Thread 18 (Thread 0x7fffa1b11700 (LWP 22712)):
#0  0x00007ffff56d712d in poll () at ../sysdeps/unix/syscall-template.S:81
No locals.
#1  0x00007ffff39e3fe4 in ?? () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
No symbol table info available.
#2  0x00007ffff39e40ec in g_main_context_iteration () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
No symbol table info available.
#3  0x00007ffff63cdae7 in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#4  0x00007ffff6376252 in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#5  0x00007ffff618596c in QThread::exec() () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#6  0x00007ffff4ec0d02 in ?? () from /home/mike/Dev/dronin/tools/Qt5.5.0/5.5/gcc_64/lib/libQt5Quick.so.5
No symbol table info available.
#7  0x00007ffff618a7df in ?? () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#8  0x00007ffff5ed1182 in start_thread (arg=0x7fffa1b11700) at pthread_create.c:312
        __res = <optimized out>
        pd = 0x7fffa1b11700
        now = <optimized out>
        unwind_buf = {cancel_jmp_buf = {{jmp_buf = {140735906125568, 591285074036167023, 1, 0, 140735906126272, 140735906125568, -591121888321040017, -591306914559651473}, mask_was_saved = 0}}, priv = {pad = {0x0, 0x0, 0x0, 0x0}, data = {prev = 0x0, cleanup = 0x0, canceltype = 0}}}
        not_first_call = <optimized out>
        pagesize_m1 = <optimized out>
        sp = <optimized out>
        freesize = <optimized out>
        __PRETTY_FUNCTION__ = "start_thread"
#9  0x00007ffff56e447d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
No locals.

Thread 17 (Thread 0x7fffa2312700 (LWP 22711)):
#0  pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
No locals.
#1  0x00007ffff618b7fb in QWaitCondition::wait(QMutex*, unsigned long) () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#2  0x00007ffff4f31c92 in ?? () from /home/mike/Dev/dronin/tools/Qt5.5.0/5.5/gcc_64/lib/libQt5Quick.so.5
No symbol table info available.
#3  0x00007ffff4f320b0 in ?? () from /home/mike/Dev/dronin/tools/Qt5.5.0/5.5/gcc_64/lib/libQt5Quick.so.5
No symbol table info available.
#4  0x00007ffff618a7df in ?? () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#5  0x00007ffff5ed1182 in start_thread (arg=0x7fffa2312700) at pthread_create.c:312
        __res = <optimized out>
        pd = 0x7fffa2312700
        now = <optimized out>
        unwind_buf = {cancel_jmp_buf = {{jmp_buf = {140735914518272, 591285074036167023, 1, 0, 140735914518976, 140735914518272, -591116390226030225, -591306914559651473}, mask_was_saved = 0}}, priv = {pad = {0x0, 0x0, 0x0, 0x0}, data = {prev = 0x0, cleanup = 0x0, canceltype = 0}}}
        not_first_call = <optimized out>
        pagesize_m1 = <optimized out>
        sp = <optimized out>
        freesize = <optimized out>
        __PRETTY_FUNCTION__ = "start_thread"
#6  0x00007ffff56e447d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
No locals.

Thread 12 (Thread 0x7fffaed65700 (LWP 22706)):
#0  0x00007ffff56d712d in poll () at ../sysdeps/unix/syscall-template.S:81
No locals.
#1  0x00007ffff39e3fe4 in ?? () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
No symbol table info available.
#2  0x00007ffff39e40ec in g_main_context_iteration () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
No symbol table info available.
#3  0x00007ffff63cdae7 in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#4  0x00007ffff6376252 in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#5  0x00007ffff618596c in QThread::exec() () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#6  0x00007ffff097f765 in ?? () from /home/mike/Dev/dronin/tools/Qt5.5.0/5.5/gcc_64/lib/libQt5Qml.so.5
No symbol table info available.
#7  0x00007ffff618a7df in ?? () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#8  0x00007ffff5ed1182 in start_thread (arg=0x7fffaed65700) at pthread_create.c:312
        __res = <optimized out>
        pd = 0x7fffaed65700
        now = <optimized out>
        unwind_buf = {cancel_jmp_buf = {{jmp_buf = {140736126670592, 591285074036167023, 1, 0, 140736126671296, 140736126670592, -591107309054553745, -591306914559651473}, mask_was_saved = 0}}, priv = {pad = {0x0, 0x0, 0x0, 0x0}, data = {prev = 0x0, cleanup = 0x0, canceltype = 0}}}
        not_first_call = <optimized out>
        pagesize_m1 = <optimized out>
        sp = <optimized out>
        freesize = <optimized out>
        __PRETTY_FUNCTION__ = "start_thread"
#9  0x00007ffff56e447d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
No locals.

Thread 11 (Thread 0x7fffc1859700 (LWP 22705)):
#0  0x00007ffff56d712d in poll () at ../sysdeps/unix/syscall-template.S:81
No locals.
#1  0x00007ffff39e3fe4 in ?? () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
No symbol table info available.
#2  0x00007ffff39e40ec in g_main_context_iteration () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
No symbol table info available.
#3  0x00007ffff63cdae7 in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#4  0x00007ffff6376252 in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#5  0x00007ffff618596c in QThread::exec() () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#6  0x00007ffff097f765 in ?? () from /home/mike/Dev/dronin/tools/Qt5.5.0/5.5/gcc_64/lib/libQt5Qml.so.5
No symbol table info available.
#7  0x00007ffff618a7df in ?? () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#8  0x00007ffff5ed1182 in start_thread (arg=0x7fffc1859700) at pthread_create.c:312
        __res = <optimized out>
        pd = 0x7fffc1859700
        now = <optimized out>
        unwind_buf = {cancel_jmp_buf = {{jmp_buf = {140736440145664, 591285074036167023, 1, 0, 140736440146368, 140736440145664, -591333093337820817, -591306914559651473}, mask_was_saved = 0}}, priv = {pad = {0x0, 0x0, 0x0, 0x0}, data = {prev = 0x0, cleanup = 0x0, canceltype = 0}}}
        not_first_call = <optimized out>
        pagesize_m1 = <optimized out>
        sp = <optimized out>
        freesize = <optimized out>
        __PRETTY_FUNCTION__ = "start_thread"
#9  0x00007ffff56e447d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
No locals.

Thread 9 (Thread 0x7fffc2ca8700 (LWP 22702)):
#0  0x00007ffff56d712d in poll () at ../sysdeps/unix/syscall-template.S:81
No locals.
#1  0x00007ffff39e3fe4 in ?? () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
No symbol table info available.
#2  0x00007ffff39e40ec in g_main_context_iteration () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
No symbol table info available.
#3  0x00007ffff63cdae7 in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#4  0x00007ffff6376252 in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#5  0x00007ffff618596c in QThread::exec() () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#6  0x00007ffff618a7df in ?? () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#7  0x00007ffff5ed1182 in start_thread (arg=0x7fffc2ca8700) at pthread_create.c:312
        __res = <optimized out>
        pd = 0x7fffc2ca8700
        now = <optimized out>
        unwind_buf = {cancel_jmp_buf = {{jmp_buf = {140736461440768, 591285074036167023, 1, 0, 140736461441472, 140736461440768, -591327040118288017, -591306914559651473}, mask_was_saved = 0}}, priv = {pad = {0x0, 0x0, 0x0, 0x0}, data = {prev = 0x0, cleanup = 0x0, canceltype = 0}}}
        not_first_call = <optimized out>
        pagesize_m1 = <optimized out>
        sp = <optimized out>
        freesize = <optimized out>
        __PRETTY_FUNCTION__ = "start_thread"
#8  0x00007ffff56e447d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
No locals.

Thread 8 (Thread 0x7fffc3bb6700 (LWP 22701)):
#0  0x00007ffff56d712d in poll () at ../sysdeps/unix/syscall-template.S:81
No locals.
#1  0x00007ffff39e3fe4 in ?? () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
No symbol table info available.
#2  0x00007ffff39e40ec in g_main_context_iteration () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
No symbol table info available.
#3  0x00007ffff63cdae7 in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#4  0x00007ffff6376252 in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#5  0x00007ffff618596c in QThread::exec() () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#6  0x00007fffc3be4235 in ?? () from /home/mike/Dev/dronin/build/ground/gcs/lib/dr/qml/QtQuick/XmlListModel/libqmlxmllistmodelplugin.so
No symbol table info available.
#7  0x00007ffff618a7df in ?? () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#8  0x00007ffff5ed1182 in start_thread (arg=0x7fffc3bb6700) at pthread_create.c:312
        __res = <optimized out>
        pd = 0x7fffc3bb6700
        now = <optimized out>
        unwind_buf = {cancel_jmp_buf = {{jmp_buf = {140736477226752, 591285074036167023, 1, 0, 140736477227456, 140736477226752, -591328542283099793, -591306914559651473}, mask_was_saved = 0}}, priv = {pad = {0x0, 0x0, 0x0, 0x0}, data = {prev = 0x0, cleanup = 0x0, canceltype = 0}}}
        not_first_call = <optimized out>
        pagesize_m1 = <optimized out>
        sp = <optimized out>
        freesize = <optimized out>
        __PRETTY_FUNCTION__ = "start_thread"
#9  0x00007ffff56e447d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
No locals.

Thread 7 (Thread 0x7fffd10f9700 (LWP 22700)):
#0  0x00007ffff56d712d in poll () at ../sysdeps/unix/syscall-template.S:81
No locals.
#1  0x00007ffff39e3fe4 in ?? () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
No symbol table info available.
#2  0x00007ffff39e40ec in g_main_context_iteration () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
No symbol table info available.
#3  0x00007ffff63cdae7 in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#4  0x00007ffff6376252 in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#5  0x00007ffff618596c in QThread::exec() () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#6  0x00007ffff097f765 in ?? () from /home/mike/Dev/dronin/tools/Qt5.5.0/5.5/gcc_64/lib/libQt5Qml.so.5
No symbol table info available.
#7  0x00007ffff618a7df in ?? () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#8  0x00007ffff5ed1182 in start_thread (arg=0x7fffd10f9700) at pthread_create.c:312
        __res = <optimized out>
        pd = 0x7fffd10f9700
        now = <optimized out>
        unwind_buf = {cancel_jmp_buf = {{jmp_buf = {140736700847872, 591285074036167023, 1, 0, 140736700848576, 140736700847872, -591369325681929873, -591306914559651473}, mask_was_saved = 0}}, priv = {pad = {0x0, 0x0, 0x0, 0x0}, data = {prev = 0x0, cleanup = 0x0, canceltype = 0}}}
        not_first_call = <optimized out>
        pagesize_m1 = <optimized out>
        sp = <optimized out>
        freesize = <optimized out>
        __PRETTY_FUNCTION__ = "start_thread"
#9  0x00007ffff56e447d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
No locals.

Thread 6 (Thread 0x7fffd1c6e700 (LWP 22698)):
#0  0x00007ffff5ed8b9d in nanosleep () at ../sysdeps/unix/syscall-template.S:81
No locals.
#1  0x00007ffff623849d in ?? () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#2  0x00007ffff6189c58 in QThread::msleep(unsigned long) () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#3  0x00007fffd4b05473 in SerialEnumerationThread::run (this=0xcd3200) at /home/mike/Dev/dronin/ground/gcs/src/plugins/serialconnection/serialplugin.cpp:81
        devices = {<QListSpecialMethods<Core::IDevice*>> = {<No data fields>}, {p = {static shared_null = {ref = {atomic = {_q_value = -1}}, alloc = 0, begin = 0, end = 0, array = {0x0}}, d = 0x7fffc8006ad0}, d = 0x7fffc8006ad0}}
#4  0x00007ffff618a7df in ?? () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#5  0x00007ffff5ed1182 in start_thread (arg=0x7fffd1c6e700) at pthread_create.c:312
        __res = <optimized out>
        pd = 0x7fffd1c6e700
        now = <optimized out>
        unwind_buf = {cancel_jmp_buf = {{jmp_buf = {140736712861440, 591285074036167023, 1, 0, 140736712862144, 140736712861440, -591368850551172753, -591306914559651473}, mask_was_saved = 0}}, priv = {pad = {0x0, 0x0, 0x0, 0x0}, data = {prev = 0x0, cleanup = 0x0, canceltype = 0}}}
        not_first_call = <optimized out>
        pagesize_m1 = <optimized out>
        sp = <optimized out>
        freesize = <optimized out>
        __PRETTY_FUNCTION__ = "start_thread"
#6  0x00007ffff56e447d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
No locals.

Thread 5 (Thread 0x7fffd246f700 (LWP 22697)):
#0  0x00007ffff56d712d in poll () at ../sysdeps/unix/syscall-template.S:81
No locals.
#1  0x00007ffff39e3fe4 in ?? () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
No symbol table info available.
#2  0x00007ffff39e40ec in g_main_context_iteration () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
No symbol table info available.
#3  0x00007ffff63cdafc in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#4  0x00007ffff6376252 in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#5  0x00007ffff618596c in QThread::exec() () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#6  0x00007ffff618a7df in ?? () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#7  0x00007ffff5ed1182 in start_thread (arg=0x7fffd246f700) at pthread_create.c:312
        __res = <optimized out>
        pd = 0x7fffd246f700
        now = <optimized out>
        unwind_buf = {cancel_jmp_buf = {{jmp_buf = {140736721254144, 591285074036167023, 1, 0, 140736721254848, 140736721254144, -591363354603646609, -591306914559651473}, mask_was_saved = 0}}, priv = {pad = {0x0, 0x0, 0x0, 0x0}, data = {prev = 0x0, cleanup = 0x0, canceltype = 0}}}
        not_first_call = <optimized out>
        pagesize_m1 = <optimized out>
        sp = <optimized out>
        freesize = <optimized out>
        __PRETTY_FUNCTION__ = "start_thread"
#8  0x00007ffff56e447d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
No locals.

Thread 4 (Thread 0x7fffd2e7b700 (LWP 22696)):
#0  0x00007ffff56d712d in poll () at ../sysdeps/unix/syscall-template.S:81
No locals.
#1  0x00007ffff39e3fe4 in ?? () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
No symbol table info available.
#2  0x00007ffff39e40ec in g_main_context_iteration () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
No symbol table info available.
#3  0x00007ffff63cdafc in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#4  0x00007ffff6376252 in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#5  0x00007ffff618596c in QThread::exec() () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#6  0x00007ffff618a7df in ?? () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#7  0x00007ffff5ed1182 in start_thread (arg=0x7fffd2e7b700) at pthread_create.c:312
        __res = <optimized out>
        pd = 0x7fffd2e7b700
        now = <optimized out>
        unwind_buf = {cancel_jmp_buf = {{jmp_buf = {140736731789056, 591285074036167023, 1, 0, 140736731789760, 140736731789056, -591361973771660945, -591306914559651473}, mask_was_saved = 0}}, priv = {pad = {0x0, 0x0, 0x0, 0x0}, data = {prev = 0x0, cleanup = 0x0, canceltype = 0}}}
        not_first_call = <optimized out>
        pagesize_m1 = <optimized out>
        sp = <optimized out>
        freesize = <optimized out>
        __PRETTY_FUNCTION__ = "start_thread"
#8  0x00007ffff56e447d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
No locals.

Thread 3 (Thread 0x7fffe3051700 (LWP 22695)):
#0  0x00007ffff56d712d in poll () at ../sysdeps/unix/syscall-template.S:81
No locals.
#1  0x00007ffff39e3fe4 in ?? () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
No symbol table info available.
#2  0x00007ffff39e40ec in g_main_context_iteration () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
No symbol table info available.
#3  0x00007ffff63cdafc in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#4  0x00007ffff6376252 in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#5  0x00007ffff618596c in QThread::exec() () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#6  0x00007ffff618a7df in ?? () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#7  0x00007ffff5ed1182 in start_thread (arg=0x7fffe3051700) at pthread_create.c:312
        __res = <optimized out>
        pd = 0x7fffe3051700
        now = <optimized out>
        unwind_buf = {cancel_jmp_buf = {{jmp_buf = {140737002149632, 591285074036167023, 1, 0, 140737002150336, 140737002149632, -591259430353727121, -591306914559651473}, mask_was_saved = 0}}, priv = {pad = {0x0, 0x0, 0x0, 0x0}, data = {prev = 0x0, cleanup = 0x0, canceltype = 0}}}
        not_first_call = <optimized out>
        pagesize_m1 = <optimized out>
        sp = <optimized out>
        freesize = <optimized out>
        __PRETTY_FUNCTION__ = "start_thread"
#8  0x00007ffff56e447d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
No locals.

Thread 2 (Thread 0x7fffea953700 (LWP 22694)):
#0  0x00007ffff56d712d in poll () at ../sysdeps/unix/syscall-template.S:81
No locals.
#1  0x00007ffff0050b72 in ?? () from /usr/lib/x86_64-linux-gnu/libxcb.so.1
No symbol table info available.
#2  0x00007ffff005264f in xcb_wait_for_event () from /usr/lib/x86_64-linux-gnu/libxcb.so.1
No symbol table info available.
#3  0x00007fffec3ae629 in ?? () from /home/mike/Dev/dronin/tools/Qt5.5.0/5.5/gcc_64/lib/libQt5XcbQpa.so.5
No symbol table info available.
#4  0x00007ffff618a7df in ?? () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#5  0x00007ffff5ed1182 in start_thread (arg=0x7fffea953700) at pthread_create.c:312
        __res = <optimized out>
        pd = 0x7fffea953700
        now = <optimized out>
        unwind_buf = {cancel_jmp_buf = {{jmp_buf = {140737129035520, 591285074036167023, 1, 0, 140737129036224, 140737129035520, -591238678145494673, -591306914559651473}, mask_was_saved = 0}}, priv = {pad = {0x0, 0x0, 0x0, 0x0}, data = {prev = 0x0, cleanup = 0x0, canceltype = 0}}}
        not_first_call = <optimized out>
        pagesize_m1 = <optimized out>
        sp = <optimized out>
        freesize = <optimized out>
        __PRETTY_FUNCTION__ = "start_thread"
#6  0x00007ffff56e447d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
No locals.

Thread 1 (Thread 0x7ffff7fbb7c0 (LWP 22690)):
#0  __lll_lock_wait () at ../nptl/sysdeps/unix/sysv/linux/x86_64/lowlevellock.S:135
No locals.
#1  0x00007ffff5ed5dc0 in pthread_cond_broadcast@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_broadcast.S:136
No locals.
#2  0x00007fffd59e688f in libusb_unlock_events () from /lib/x86_64-linux-gnu/libusb-1.0.so.0
No symbol table info available.
#3  0x00007fffd59e2db6 in ?? () from /lib/x86_64-linux-gnu/libusb-1.0.so.0
No symbol table info available.
#4  0x00007fffd59e2fb8 in libusb_close () from /lib/x86_64-linux-gnu/libusb-1.0.so.0
No symbol table info available.
#5  0x00007fffd5c045a0 in hid_close (dev=0x49d7200) at /home/mike/Dev/dronin/ground/gcs/src/plugins/rawhid/hidapi/hidapi_linux.c:1255
No locals.
#6  0x00007fffd4f6b68a in tl_dfu::DFUObject::CloseBootloaderComs (this=0x44b3a60) at /home/mike/Dev/dronin/ground/gcs/src/plugins/uploader/tl_dfu.cpp:450
No locals.
#7  0x00007fffd4f6b2dc in tl_dfu::DFUObject::OpenBootloaderComs (this=0x44b3a60, port=...) at /home/mike/Dev/dronin/ground/gcs/src/plugins/uploader/tl_dfu.cpp:409
        m_eventloop = <incomplete type>
        __PRETTY_FUNCTION__ = "bool tl_dfu::DFUObject::OpenBootloaderComs(USBPortInfo)"
#8  0x00007fffd4f4b88c in uploader::UploaderGadgetWidget::onBootloaderDetected (this=0x44b3a10) at /home/mike/Dev/dronin/ground/gcs/src/plugins/uploader/uploadergadgetwidget.cpp:466
        brdMgr = 0x9afe90
        devices = {<QListSpecialMethods<USBPortInfo>> = {<No data fields>}, {p = {static shared_null = {ref = {atomic = {_q_value = -1}}, alloc = 0, begin = 0, end = 0, array = {0x0}}, d = 0x7d53b60}, d = 0x7d53b60}}
#9  0x00007fffd4f6eccd in uploader::UploaderGadgetWidget::qt_static_metacall (_o=0x44b3a10, _c=QMetaObject::InvokeMetaMethod, _id=17, _a=0x7fffffffd090) at .moc/debug-shared/moc_uploadergadgetwidget.cpp:223
        _t = 0x44b3a10
#10 0x00007ffff63a636e in QMetaObject::activate(QObject*, int, int, void**) () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#11 0x00007fffd5c06825 in USBSignalFilter::deviceDiscovered (this=0x378e610) at .moc/debug-shared/moc_usbsignalfilter.cpp:149
No locals.
#12 0x00007fffd5c01ee1 in USBSignalFilter::m_deviceDiscovered (this=0x378e610, port=...) at /home/mike/Dev/dronin/ground/gcs/src/plugins/rawhid/usbsignalfilter.cpp:10
        __PRETTY_FUNCTION__ = "void USBSignalFilter::m_deviceDiscovered(USBPortInfo)"
#13 0x00007fffd5c065c6 in USBSignalFilter::qt_static_metacall (_o=0x378e610, _c=QMetaObject::InvokeMetaMethod, _id=2, _a=0x7fffffffd350) at .moc/debug-shared/moc_usbsignalfilter.cpp:88
        _t = 0x378e610
#14 0x00007ffff63a636e in QMetaObject::activate(QObject*, int, int, void**) () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#15 0x00007fffd5c064df in USBMonitor::deviceDiscovered (this=0xcd6950, _t1=...) at .moc/debug-shared/moc_usbmonitor.cpp:145
        _a = {0x0, 0x7fffffffd3a0}
#16 0x00007fffd5c04c5b in USBMonitor::deviceEventReceived (this=0xcd6950) at /home/mike/Dev/dronin/ground/gcs/src/plugins/rawhid/usbmonitor_linux.cpp:77
        action = {static null = {<No data fields>}, d = 0x569c390}
        devtype = {static null = {<No data fields>}, d = 0x2f0cd90}
        __PRETTY_FUNCTION__ = "void USBMonitor::deviceEventReceived()"
        dev = 0x5653880
#17 0x00007fffd5c062e4 in USBMonitor::qt_static_metacall (_o=0xcd6950, _c=QMetaObject::InvokeMetaMethod, _id=2, _a=0x7fffffffd580) at .moc/debug-shared/moc_usbmonitor.cpp:84
        _t = 0xcd6950
#18 0x00007ffff63a636e in QMetaObject::activate(QObject*, int, int, void**) () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#19 0x00007ffff642546e in QSocketNotifier::activated(int, QSocketNotifier::QPrivateSignal) () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#20 0x00007ffff63b34eb in QSocketNotifier::event(QEvent*) () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#21 0x00007ffff71944dc in QApplicationPrivate::notify_helper(QObject*, QEvent*) () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Widgets.so.5
No symbol table info available.
#22 0x00007ffff7199640 in QApplication::notify(QObject*, QEvent*) () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Widgets.so.5
No symbol table info available.
#23 0x00007ffff63786a3 in QCoreApplication::notifyInternal(QObject*, QEvent*) () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#24 0x00007ffff63ce045 in ?? () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#25 0x00007ffff39e3e04 in g_main_context_dispatch () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
No symbol table info available.
#26 0x00007ffff39e4048 in ?? () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
No symbol table info available.
#27 0x00007ffff39e40ec in g_main_context_iteration () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
No symbol table info available.
#28 0x00007ffff63cdae7 in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#29 0x00007ffff6376252 in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#30 0x00007ffff637dd3d in QCoreApplication::exec() () from /home/mike/Dev/dronin/build/ground/gcs/bin/../lib/dr/libQt5Core.so.5
No symbol table info available.
#31 0x000000000040e51b in main (argc=1, argv=0x7fffffffe398) at /home/mike/Dev/dronin/ground/gcs/src/app/main.cpp:390
        settingsFilename = {static null = {<No data fields>}, d = 0x6d46d0}
        positionalArguments = {<QList<QString>> = {<QListSpecialMethods<QString>> = {<No data fields>}, {p = {static shared_null = {ref = {atomic = {_q_value = -1}}, alloc = 0, begin = 0, end = 0, array = {0x0}}, d = 0x7ffff6468a30 <QListData::shared_null>}, d = 0x7ffff6468a30 <QListData::shared_null>}}, <No data fields>}
        gcsTranslationsPath = @0x7fffffffded0: {static null = {<No data fields>}, d = 0x7b0a10}
        parser = {d = 0x6d3430}
        dirName = {static null = {<No data fields>}, d = 0x6d0f80}
        locale = {static null = {<No data fields>}, d = 0x6d5e30}
        versionOption = {d = {d = 0x6d3890}}
        qtTranslator = <incomplete type>
        parsingErrors = {<QList<QString>> = {<QListSpecialMethods<QString>> = {<No data fields>}, {p = {static shared_null = {ref = {atomic = {_q_value = -1}}, alloc = 0, begin = 0, end = 0, array = {0x0}}, d = 0x7ffff6468a30 <QListData::shared_null>}, d = 0x7ffff6468a30 <QListData::shared_null>}}, <No data fields>}
        plugins = {<QListSpecialMethods<ExtensionSystem::PluginSpec*>> = {<No data fields>}, {p = {static shared_null = {ref = {atomic = {_q_value = -1}}, alloc = 0, begin = 0, end = 0, array = {0x0}}, d = 0x8ebc80}, d = 0x8ebc80}}
        coreplugin = 0x7cd060
        helpOption = {d = {d = 0x6d31d0}}
        overrideOption = {d = {d = 0x6d3b50}}
        noLoadOption = {d = {d = 0x6d3ef0}}
        pluginOption = {d = {d = 0x6d4350}}
        translator = <incomplete type>
        pixmap = <incomplete type>
        app = <incomplete type>
        showVersionOption = {d = {d = 0x6d39a0}}
        cleanConfig = {d = {d = 0x6d3d00}}
        doTestsOption = {d = {d = 0x6d4160}}
        splash = {<QSplashScreen> = {<No data fields>}, static staticMetaObject = {d = {superdata = 0x7ffff78b0ee0 <QSplashScreen::staticMetaObject>, stringdata = 0x433cc0 <qt_meta_stringdata_CustomSplash>, data = 0x433da0 <qt_meta_data_CustomSplash>, static_metacall = 0x41275e <CustomSplash::qt_static_metacall(QObject*, QMetaObject::Call, int, void**)>, relatedMetaObjects = 0x0, extradata = 0x0}}, m_progress = 100, m_pixmap = <incomplete type>, m_progress_bar_color = {cspec = QColor::Rgb, ct = {argb = {alpha = 65535, red = 55769, green = 30326, blue = 0, pad = 0}, ahsv = {alpha = 65535, hue = 55769, saturation = 30326, value = 0, pad = 0}, acmyk = {alpha = 65535, cyan = 55769, magenta = 30326, yellow = 0, black = 0}, ahsl = {alpha = 65535, hue = 55769, saturation = 30326, lightness = 0, pad = 0}, array = {65535, 55769, 30326, 0, 0}}}, time = {mds = 10388622}, message_number = 213, settings = <incomplete type>}
        pluginManager = {<QObject> = {<No data fields>}, static staticMetaObject = {d = {superdata = 0x7ffff681b300 <QObject::staticMetaObject>, stringdata = 0x7ffff78f8460 <qt_meta_stringdata_ExtensionSystem__PluginManager>, data = 0x7ffff78f8600 <qt_meta_data_ExtensionSystem__PluginManager>, static_metacall = 0x7ffff78f4fc4 <ExtensionSystem::PluginManager::qt_static_metacall(QObject*, QMetaObject::Call, int, void**)>, relatedMetaObjects = 0x0, extradata = 0x0}}, d = 0x765810, static m_instance = 0x7fffffffe1d0, m_lock = {d = 0x765b60}, m_allPluginsLoaded = true}
        pluginPaths = {<QList<QString>> = {<QListSpecialMethods<QString>> = {<No data fields>}, {p = {static shared_null = {ref = {atomic = {_q_value = -1}}, alloc = 0, begin = 0, end = 0, array = {0x0}}, d = 0x6d8550}, d = 0x6d8550}}, <No data fields>}
```

tracernz commented:
Okay, got a SIGABRT in the same place this time. https://gist.github.com/tracernz/6b7423115c24de4f9d07

mlyle commented:
For clarity, what do you mean by hang-- does the whole UI lock up?

tracernz commented:
Yes, whole UI hangs most times when the BL starts (e.g. shortly after clicking halt, or when plugging the board in. Something like this https://imgur.com/gallery/veAyHcP. It never recovered in the couple of minutes or so I allowed it.

mlyle committed:
uploader: dfu: simplify open state / handle management

mlyle commented:
OK, that's interesting.  I've had problems like that a fair bit without this changeset, and oddly enough this changeset seemed to have made them better.  I've made one minor change to the uploader dfu code that I don't really expect to help things.  I don't have a Linux box running GCS presently but I can certainly play with CC3D on OSX.

mlyle commented:
This indeed does break for me-- immediate crash-- for CC3D on OSX in rescue.  interesting.

mlyle committed:
uploader: properly init m_hidHandle

mlyle commented:
OK, so my testing was complicated because even in the previous state the repro was not reliable.  Maybe 50-50 on the first rescue, and then really reliable for the rest of the session.

This lead me to think maybe it's the lack of initializing m_hidHandle that was guilty (duh).   I fixed that and it has been reliable for me for 8 GCS starts + immediate rescues plus other test sequences.

@tracernz can you please try, and see if this is better for you on Linux?

tracernz commented:
SIGBUS this time https://gist.github.com/tracernz/9dab5641a231ac8c5c3e. Repro is start GCS, plug CC3D in, wait for telem to connect, click halt. Once the bootloader starts boom. It also happens most times if you starts gcs, connect board, unplug board, reconnect board.

mlyle commented:
jenkins, build artifacts please

mlyle commented:
@tracernz --- thanks for finding this breakage.

I'm still mystified.  But I have coerced my VM machine to run the product and have repro'd on CC3D.

mlyle committed:
gcs: hid/uploader: further simplify things

Remove portions of lock model and state machine that are not needed and
interfere with robustness.  Fix shutdown model in hidapi_linux.

mlyle commented:
Did several cycles on Linux with this new changeset successfully.  Now I need to bang on OSX a little to confirm I've made things no worse.

mlyle commented:
Needs further work.  This combination is regressed on OSX.

mlyle commented:
"More than one device detected in bootloader state"--- which has been a relevant / occasional failure elsewhere.  So now it's reliably manifested, so that's good.

mlyle commented:
reliable with revo mini.. interestink.

mlyle commented:
OK, I'm super dumb, I had 2 flight boards plugged into the machine I was testing CC3D on.

I've gone through a long sequence of tests of different combinations (start GCS with board connected, without, rescue, halt, etc) on OSX and it's been reliable across revomini and CC3D.

Still, additional test would be very much appreciated since these are fiddly bits.

mlyle commented:
jenkins build artifacts please

mlyle commented:
gonna play with Windows some next (perhaps later today).

mlyle commented:
jenkins, build artifacts please

mlyle commented:
jenkins, test this please

mlyle commented:
jenkins, test this please

mlyle commented:
jenkins, build artifacts please

mlyle commented:
jenkins, test this please

mlyle commented:
jenkins build artifacts please

mlyle committed:
gcs: take less of forever to connect.

Also update the connection status object more frequently so we're less
likely to spuriously lose connection from little blips / lost messages.

ufoDziner commented:
Tested the portable version on Win10 with Brain FPV.
Takes 6sec to connect instead of 12 with the release with the board plugged in after the GCS is running.
With the board plugged in prior, connection is immediate.
Rescue worked the same.

tracernz commented:
Doesn't crash now on Linux :D. The CC3D still doesn't halt into bootloader (which is the same as without this PR) but it does rescue okay and telemetry connects/disconnects/unplugs/replugs all seem to be fine.

mlyle commented:
Definitely on OSX it -can- get confused complaining about multiple boards.  I have a feeling most of these cases would have crashed before-- now it's rock solid and not crashy, but occasionally complainy.  There is room for further progress but I don't see this as a showstopper.

mlyle commented:
This isn't ready.

Windows can be convinced to be crashy with this... and on both Windows and Mac I have seen the first attempt to click "flash" fail, with the subsequent one succeeding.

mlyle commented:
@ekaszubski reports in #582 that he gets this same behavior with current next (first "flash" click failing, next working).

ufoDziner commented:
Is this to say that there are sounds that imply that it's connected, then disconnected, before you get the sounds and visuals that show it connecting?  If so, I get a false first connect as wess.

mlyle commented:
No.  @Eka reports a regression on mainline where clicking 'flash' returns an error.  Clicking 'flash' in the bootloader again works.

That regression is what had caused me to abandon this branch for awhile.

mlyle commented:
Not happy at the moment brain + mac.  More thought needed.

mlyle commented:
Scenario 1: OSX, Lux.  Rescue, boot, halt, normal plug in, etc.  Very happy and snappy.
Scenario 2: OSX, Brain.  Rescue, boot, halt, normal plug in.  Very happy and snappy (noticeably less snappy than Lux, though-- maybe the additional objects or some other factor)

These two are definitely not less reliable than baseline.

mlyle commented:
rebased and force-push'd.

jenkins, build artifacts please

mlyle commented:
Scenario 3: brain, windows.  Rescue, boot, halt, normal plugin.  Halt results in "Trying to connect bootloader" message sticking around when bootloader is actually connected and functional.   After extended testing finally got a hang, but I think this is comparable to baseline.

Scenario 4: lux, windows.  Rescue, boot, halt, normal plug.  Works great.

Scenario 5: cc3d, windows.  Rescue, boot, halt, normal plug.  Works great.

Scenario 6: quanton, windows.  Rescue, boot, halt, normal plug.  Works great.

mlyle commented:
Scenario 7: quanton, linux.  Rescue, boot, halt, normal plug.  Works great.
Scenario 8: oplm, linux. Rescue, boot, normal plug.  Works great.
Scenario 9: brain, linux.  Rescue, boot, halt, normal plug.  Works great.  No bootloader message anomaly.
Scenario 10: lux, Linux.  Rescue, boot, halt, normal plug.  Works great.

mlyle commented:
Recommend other people test this and we merge it soon.  It isn't quite perfect.  But it seems to be considerably better than what we had before on my systems.

The main anomaly that caused me to abandon this before (CC3D not flashing) is not related to this changeset and is in next.

ufoDziner commented:
Sparky2 tested well.  Only had one instance that the Brain wouldn't boot after pressing the reboot button.  I connected it, entered bootloader, boot, then reboot, then it hung.  Couldn't recreate it, so could have been me.  Overall it seems more clear.  Very fast reboots as well.  I would definitely use this more often (I normally plug/unplug the cable).

mlyle commented:
@ufoDziner nice-- much appreciated.

(The UI changes are from an earlier @tracernz PR, but this makes them all snappier)

tracernz commented:
CC3D in bootloader mode causes GCS to hang every time. If the CC3D is sitting in bootloader mode, and I start GCS, it hangs on the splash screen while loading the Uploader plugin. If GCS is already running it hangs as soon as I connect the CC3D. x64 Ubuntu 14.04 etc.

tracernz commented:
Main thread hangs here https://github.com/mlyle/dronin/blob/mpl-connectfaster/ground/gcs/src/plugins/rawhid/hidapi/hidapi_linux.c#L1177
The read_thread is still running.

mlyle commented:
A new `thread apply all bt`   from gdb would be really welcome (so I can see where / why the read_thread is hung, and what any libusb threads are doing).

I've spent about a half hour trying to reproduce this--- things I've tried (all CC3D + linux):
- Normal rescue 5-6 times
- 'Enter bootloader'/halt 5-6 times
- Rescue, quit GCS, re-enter GCS (results in about a 1.5 seconds delay while opening bootloader 2-3 times
- Switching to XPRA instead of remote X and running the same tests
- Flashing all 1's to the application area -- running and then starting GCS, 3-4 times.  Results in "failed to open connection to bootloader" but no crash / hang.
- Fresh GCS launches and many of the above scenarios.

`Linux squanch 4.2.0-16-generic #19-Ubuntu SMP Thu Oct 8 15:35:06 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux`

mlyle committed:
gcs: hid: make everything everywhere always happy

tracernz commented:
Fixed, yay! :D

mlyle committed:
gcs: uploader/dfu: prevent invalid states

tracernz committed:
GCS:DFU: Clean up debug messages, handle additional status

Additional status gives the last command which is useful if it failed

tracernz committed:
GCS:DFU: Fix F1 bootloader so first flash succeeds

tracernz committed:
GCS:DFU: Fix safe boot on older F1 bootloader

Also fixed in bootloader in #740

mlyle committed:
Merge pull request #3 from tracernz/mpl-connectfaster

Fix cc3d flashing, document some stuff, clean up debug messages in dfu, fix safe boot

mlyle commented:
Tested the new commits from @tracernz's PRs-- seems good.  Flash works on the first try on CC, safeboot works.
